vaadin-testbench-core/src/main/kotlin/com/vaadin/testbench/unit/internal/BasicUtils.kt

```diff
@@ -0,0 +1,178 @@
+        else -> isVisible
+    }
+
+/**
+ * Returns direct text contents (it doesn't peek into the child elements).
+ */
+val Component._text: String?
+    get() = when (this) {
+        is HasText -> text
+        is Text -> text   // workaround for https://github.com/vaadin/flow/issues/3606
+        else -> null
+    }
+
+/**
+ * Checks that a component is actually editable by the user:
+ * * The component must be effectively visible: it itself must be visible, its parent must be visible and all of its ascendants must be visible.
+ *   For the purpose of testing individual components not attached to the [UI], a component may be considered visible even though it's not
+ *   currently nested in a [UI].
+ * * The component must be effectively enabled: it itself must be enabled, its parent must be enabled and all of its ascendants must be enabled.
+ * * If the component is [HasValue], it must not be [HasValue.isReadOnly].
+ * @throws IllegalStateException if any of the above doesn't hold.
+ */
+fun Component.checkEditableByUser() {
+    check(isEffectivelyVisible()) { "The ${toPrettyString()} is not effectively visible - either it is hidden, or its ascendant is hidden" }
+    val parentNullOrEnabled = !parent.isPresent || parent.get().isEffectivelyEnabled()
+    if (parentNullOrEnabled) {
+        check(isEnabled) { "The ${toPrettyString()} is not enabled" }
+    }
+    check(isEffectivelyEnabled()) { "The ${toPrettyString()} is nested in a disabled component" }
+    if (this is HasValue<*, *>) {
+        @Suppress("UNCHECKED_CAST")
+        val hasValue = this as HasValue<HasValue.ValueChangeEvent<Any?>, Any?>
+        check(!hasValue.isReadOnly) { "The ${toPrettyString()} is read-only" }
+    }
+}
+
+/**
+ * Fails if the component is editable. See [checkEditableByUser] for more details.
+ * @throws AssertionError if the component is editable.
+ */
+fun Component.expectNotEditableByUser() {
+    try {
+        checkEditableByUser()
+        throw AssertionError("The ${toPrettyString()} is editable")
+    } catch (ex: IllegalStateException) {
+        // okay
+    }
+}
+
+internal fun Component.isEffectivelyVisible(): Boolean = _isVisible && (!parent.isPresent || parent.get().isEffectivelyVisible())
+
+/**
+ * Computes whether this component and all of its parents are enabled.
+ *
+ * Effectively a shortcut for [isEnabled] since it recursively checks that all ancestors
+ * are also enabled (the "implicitly disabled" effect, see [HasEnabled.isEnabled] javadoc for more details).
+ * @return false if this component or any of its parent is disabled.
+ */
+fun Component.isEffectivelyEnabled(): Boolean = isEnabled
+
+/**
+ * Checks whether this component is [HasEnabled.isEnabled]. All components not implementing [HasEnabled] are considered enabled.
+ */
+val Component.isEnabled: Boolean
+    get() = when (this) {
+        is HasEnabled -> isEnabled
+        else -> true
+    }
```

Comment:
This is totally skiping the fact that the check should be `component.getElement().isEnabled()` which is basically what the mixin gets as a shorthand.
This should be removed as it makes things wrong as a non HasEnabled that is in a disabled component should return that it is disabled.

Reply:
added ticket #1354